vw/src/main/scala/com/microsoft/azure/synapse/ml/vw/VowpalWabbitCSETransformer.scala

```diff
@@ -0,0 +1,208 @@
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.vw
+
+import com.microsoft.azure.synapse.ml.codegen.Wrappable
+import com.microsoft.azure.synapse.ml.logging.BasicLogging
+import com.microsoft.azure.synapse.ml.policyeval.PolicyEvalUDAFUtil
+import org.apache.spark.ml.param.{FloatParam, ParamMap, StringArrayParam}
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.ml.{ComplexParamsWritable, Transformer}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{Column, DataFrame, Dataset, functions => F, types => T}
+
+/**
+  * Emits continuous success ev? metrics for contextual bandit style predictions and logs.
+  */
+class VowpalWabbitCSETransformer(override val uid: String)
+  extends Transformer
+    with BasicLogging
+    with Wrappable
+    with ComplexParamsWritable {
+
+  import VowpalWabbitDSJsonTransformer._
+  import VowpalWabbitCSETransformer._
+
+  logClass()
+
+  def this() = this(Identifiable.randomUID("VowpalWabbitEvalTransformer"))
+
+  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)
+
+  val minImportanceWeight = new FloatParam(
+    this, "dsJsonColumn", "Cap importance weight at this lower bound. Defaults to 0.")
+
+  def getMinImportanceWeight: Float = $(minImportanceWeight)
+  def getMinImportanceWeight(value: Float): this.type = set(minImportanceWeight, value)
+
+  val maxImportanceWeight = new FloatParam(
+    this, "dsJsonColumn", "Cap importance weight at this upper bound. Defaults to 100.")
+
+  def getMaxImportanceWeight: Float = $(maxImportanceWeight)
+  def getMaxImportanceWeight(value: Float): this.type = set(maxImportanceWeight, value)
+
+  val metricsStratificationCols = new StringArrayParam(
+    this, "metricsStratificationCols", "Optional list of column names to stratify rewards by.")
+
+  def getMetricsStratificationCols: Array[String] = $(metricsStratificationCols)
+  def getMetricsStratificationCols(value: Array[String]): this.type = set(metricsStratificationCols, value)
+
+  setDefault(minImportanceWeight -> 0, maxImportanceWeight -> 100, metricsStratificationCols -> Array.empty)
+
+  // define reward independent metrics
+  val globalMetrics: Seq[Column] = Seq(
+    F.count("*").alias(ExampleCountName),
+    F.sum(F.when(F.expr(s"$ProbabilityPredictedColName > 0"), 1).otherwise(0))
+      .alias(ProbabilityPredictedNonZeroCount),
+    F.min("w").alias(MinimumImportanceWeight),
+    F.max("w").alias(MaximumImportanceWeight),
+    F.expr("avg(w)").alias(AverageImportanceWeight),
+    F.expr("avg(w * w)").alias(AverageSquaredImportanceWeight),
+    F.expr("max(w)/count(*)").alias(PropOfMaximumImportanceWeight),
+    F.expr("approx_percentile(w, array(0.25, 0.5, 0.75, 0.95))")
+      .alias(ImportanceWeightQuantiles))
+
+  case class RewardColumn(name: String, col: String, idx: Int) {
+    def minRewardCol: String = s"min_reward_$idx"
```

Comment:
perhaps if this in an indexed set of columns just make it an array? not sure how this factors into your contracts though

Reply:
I need unique column names for min & max and I need to fetch them later... not sure how to change.